### ISSUE TITLE
Add a flavor of `training_step` that allows for expressing inter-batch parallelism

### ISSUE THREAD

**Issue by yifuwang** [PyTorchLightning/pytorch-lightning#8316]
### Motivation

When training a model, the most common approach is to process input batches sequentially (Lightning does this today). This makes sense, because in order to achieve synchronous/deterministic training, the processing of batch[n+1] can’t start until the optimization for batch[n] completes. However, this is only partly true, because it only applies to model weight-dependent logic. We can execute the model weight-independent logic out-of-order while still maintaining numerical equivalence. If the model weight-independent logic takes significant amount of wall time and stresses different system resources compared to the rest of the training step, parallelization can lead to increased throughput.

A trivial example of this is hiding the latency of host-to-device copy of input batches behind computational intensive operations. Without parallization, the execution looks like:
```
batch 0: [HtoD][forward][backward]
batch 1:                          [HtoD][forward][backward]
batch 2:                                                   [HtoD][forward][backward]
```
With parallelization, the latency of HtoD copy can be hidden:
```
batch 0: [HtoD][forward][backward]
batch 1:       [HtoD]             [forward][backward]
batch 2:             [HtoD]                          [forward][backward]
```
A more concrete use case for inter-batch parallelism at FB is to accelerate the model parallel technique used in DLRM (https://github.com/facebookresearch/dlrm). DLRM models consist large amount of embedding layers which are sharded across different GPUs. The forward pass for the embedding layers is roughly as follows:
1. All ranks receive global indices
2. The indices are routed to GPUs that hold the corresponding embedding weights (with alltoall collective)
3. All ranks perform local embedding lookup
4. The embedding lookup results are routed back to senders of the indices (with alltoall collective)

As we can see, step 1 and 2 are (1) model weight-independent (2) expensive (3) stresses communication resources. With inter-batch parallelism, we can hide the latency behind the previous batch’s backward pass.

### Pitch

Support a `training_step` flavor with access to the data iterator, so users can express the aforementioned algorithms.

**Comment by ananthsub** [comment]
@yifuwang how much do you want to tie this feature support to the LightningModule decomposition proposed elsewhere?

If we want to decouple this, in the short term we could follow a similar strategy that manual optimization took:
- Introduce a new property and setter on the LightningModule e.g. `automatic_iteration` 
- Depending on the property set on the LightningModule, add conditional logic in the loops to dynamically send the right arguments to the step function (e.g. the batch or the data iterator)

Then in the long term, we continue formalizing these as separate interfaces

My assumption is that this would hold for all of `training_step`, `validation_step`, `test_step`, `predict_step` as the non-training ones would still benefit from the HtoD overlap with `forward` right?

**Comment by zzzwen** [comment]
Question, are you planning to let `training_step` take an union of a "dataframe" and an "iter of dataframe" or this will be a different function altogether?

**Comment by yifuwang** [comment]
@ananthsub I haven't considered how to approach this from LightningModule decomposition's perspective. But I think it makes perfect sense to do so.

**Comment by tchaton** [comment]
Dear @yifuwang,

Would you mind pointing us to the current implementation of such features within DLRM library or provide a pseudo-code on how this would work ?

Best,
T.C

**Comment by yifuwang** [comment]
Hi @tchaton,

Sorry for the delayed response.

It looks like the DRLM project didn't include the optimization. However, a WIP recsys project that FB may open source soon will rely on the optimization. Please see below for a simulation of the problem:

Best,
Yifu

```python3
#!/usr/bin/env python3
import torch
from statistics import mean
import time


CYCLES_PER_MS = None


def get_cycles_per_ms() -> float:
    """
    Measure and return approximate number of cycles per millisecond for torch.cuda._sleep

    Copied from: github.com/pytorch/pytorch/blob/master/test/test_cuda.py
    """

    def measure() -> float:
        start = torch.cuda.Event(enable_timing=True)
        end = torch.cuda.Event(enable_timing=True)
        start.record()
        torch.cuda._sleep(1000000)
        end.record()
        end.synchronize()
        cycles_per_ms = 1000000 / start.elapsed_time(end)
        return cycles_per_ms

    # Get 10 values and remove the 2 max and 2 min and return the avg.
    # This is to avoid system disturbance that skew the results, e.g.
    # the very first cuda call likely does a bunch of init, which takes
    # much longer than subsequent calls.
    #
    # Tested on both Tesla V100, Quadro GP100, Titan RTX, RTX 3090 GPUs
    # and seems to return stable values. Therefore, we enable caching
    # using lru_cache decorator above.
    num = 10
    vals = []
    for _ in range(num):
        vals.append(measure())
    vals = sorted(vals)
    return mean(vals[2 : num - 2])


CYCLES_PER_MS = int(get_cycles_per_ms())
BATCH_SIZE = 128
EMB_SZ = 100
EMB_DIM = 64


class Model(torch.nn.Module):
    """
    A toy model for mimicking the communication overhead of sharded embedding
    modules in DLRM models.

    DLRM models can be trained in a DDP-like fashion, where each trainer
    receives different batches (embedding indices in this example). Since the
    embeddings are sharded across trainers, the lookup process involves (1)
    routing the indices to the trainer that possesses the corresponding
    embeddings (2) performing local lookup (3) routing the embedding lookup
    result back.

    The toy model doesn't actually perform index/result routing. It simply
    uses torch.cuda._sleep() to mimic the cost of the communication op (i.e.
    a2a).
    """

    def __init__(self):
        super().__init__()
        self.local_embedding = torch.nn.Embedding(EMB_SZ, EMB_DIM)

    def forward(self, indices: torch.Tensor):
        local_indices = self._route_indices(indices)
        result = self.local_embedding(local_indices)
        return self._route_result(result)

    def _route_indices(self, batch: torch.Tensor, non_blocking=False):
        """
        This can be parallelized across different batches since it's model
        weight independent.

        Why not run this in dataloader/datamodule?
        - The routing logic depends on how model is sharded
        - Putting this in data preprocessor changes the semantic of the model
        """
        torch.cuda._sleep(CYCLES_PER_MS * 1_000)
        if not non_blocking:
            torch.cuda.synchronize()
        return batch

    def _route_result(self, result: torch.Tensor, non_blocking=False):
        torch.cuda._sleep(CYCLES_PER_MS * 1_000)
        if not non_blocking:
            torch.cuda.synchronize()
        return result


def get_dataloader():
    for _ in range(10):
        yield torch.randint(EMB_DIM, [BATCH_SIZE])


 def non_parallelized():
     dataloader = get_dataloader()
     for batch in dataloader:
         m(batch)


 def parallelized():
     dataloader = get_dataloader()

     comm_stream = torch.cuda.Stream()
     batch_i_ready = torch.cuda.Event()

     # Route the first batch and kick off the routing for the second batch
     with torch.cuda.stream(comm_stream):
         batch_i = next(dataloader)
         m._route_indices(batch_i, non_blocking=True)
         batch_i_ready.record()

     # Invariant: the routing for batch[i] has been kicked off
     done = False
     while not done:
         with torch.cuda.stream(comm_stream):
             batch_ip1_ready = torch.cuda.Event()
             try:
                 batch_ip1 = next(dataloader)
                 m._route_indices(batch_ip1, non_blocking=True)
                 batch_ip1_ready.record()
             except StopIteration:
                 done = True

         batch_i_ready.wait()

         result = m.local_embedding(batch_i)
         m._route_result(result)

         batch_i = batch_ip1
         batch_i_ready = batch_ip1_ready


 if __name__ == "__main__":
     torch.cuda.set_device(torch.device("cuda:0"))

     m = Model()

     begin_ts = time.time()
     non_parallelized()
     print(f"The non-parallelized version took {int(time.time() - begin_ts)} seconds")

     begin_ts = time.time()
     parallelized()
     print(f"The parallelized version took {int(time.time() - begin_ts)} seconds")
```

**Comment by tchaton** [comment]
Dear @yifuwang,

I have been thinking about the cleanest way to do this with Lighting. 
Mind having a look at https://github.com/PyTorchLightning/pytorch-lightning/pull/8700 ?

This line should have to be moved to LightingModule `on_after_batch_transfer` hook for this work properly. 

```py
m._route_indices(batch_ip1, non_blocking=True)
```

This is just an idea and would like to get your thoughts on this first design.

Best,
T.C

**Comment by yifuwang** [comment]
Dear @tchaton,

Thanks for looking into this and coming up with a proposal so quickly!

I really like the idea of having Lightning abstract away this type of parallelism. However, as we discussed earlier, for FB's use case, we still need the flexibility to experiment with different parallelization strategies (even tradeoff between model accuracy and training performance). I'd like to propose that we still add the flavor of `training_step` with data iterator access (maybe as an experimental API), while pursuing a more user friendly solution (e.g. #8700) for more general use cases. How does it sound?

If you are okay with this, I'd like to give `training_step` with data iterator access a shot.

Best,
Yifu

**Comment by tchaton** [comment]
Dear @yifuwang,

@kaushikb11 made some extra progress on #8700. 

The feature is currently supported only for the GPU accelerator.
We were thinking `num_prefetch_batches=-1` could used to request the iterator to be passed directly to the `training_step` as required by FB.
This assumes fetching, moving data to device (using Lightning utilities) is left to the user control.

One limitation for this approach for Lightning is that the provided `batch_idx` passed to the `training_step` can't be inferred and should be provided as None.

Feel free to collaborate with @kaushikb11 on extending Lightning to pass the iterator directly :)

Best,
T.C

**Comment by ananthsub** [comment]
I was going back and found some interesting issues previously raised in lightning which discussed this: 
- https://github.com/PyTorchLightning/pytorch-lightning/issues/2611
- https://github.com/PyTorchLightning/pytorch-lightning/issues/1316 
- https://github.com/PyTorchLightning/pytorch-lightning/pull/1521
- https://github.com/PyTorchLightning/pytorch-lightning/issues/4803
